homeassistant/components/asterisk_mbox.py

```diff
@@ -0,0 +1,192 @@
+        hass.http.register_view(AsteriskMboxMP3View(ASTERISK))
+        hass.http.register_view(AsteriskMboxDeleteView(ASTERISK))
+
+    return True
+
+
+class AsteriskData(object):
+    """Store Asterisk mailbox data."""
+
+    def __init__(self, hass, host, port, password):
+        """Init the Asterisk data object."""
+        from asterisk_mbox import Client as asteriskClient
+
+        self.hass = hass
+        self.client = asteriskClient(host, port, password, self.handle_data)
+        self.messages = []
+
+        async_dispatcher_connect(
+            self.hass, SIGNAL_MESSAGE_REQUEST, self._request_messages)
+
+    @callback
+    def handle_data(self, command, msg):
+        """Handle changes to the mailbox."""
+        from asterisk_mbox.commands import CMD_MESSAGE_LIST
+
+        if command == CMD_MESSAGE_LIST:
+            _LOGGER.info("AsteriskVM sent updated message list")
+            self.messages = sorted(msg,
+                                   key=lambda item: item['info']['origtime'],
+                                   reverse=True)
+            async_dispatcher_send(self.hass, SIGNAL_MESSAGE_UPDATE,
+                                  self.messages)
+
+    @callback
+    def _request_messages(self):
+        """Handle changes to the mailbox."""
+        _LOGGER.info("Requesting message list")
+        self.client.messages()
+
+
+class AsteriskMboxMsgView(HomeAssistantView):
+    """View to return the list of messages."""
+
+    url = "/api/asteriskmbox/messages"
+    name = "api:asteriskmbox:messages"
+
+    def __init__(self, data):
+        """Initialize."""
+        self.data = data
+
+    @asyncio.coroutine
+    def get(self, request):
+        """Retrieve Asterisk messages."""
+        msgs = self.data.messages
+        _LOGGER.info("Sending: %s", msgs)
+        return self.json(msgs)
+
+
+class AsteriskMboxDeleteView(HomeAssistantView):
+    """View to delete selected messages."""
+
+    url = "/api/asteriskmbox/delete"
+    name = "api:asteriskmbox:delete"
+
+    def __init__(self, data):
+        """Initialize."""
+        self.data = data
+
+    @asyncio.coroutine
+    def post(self, request):
+        """Delete items."""
+        try:
+            data = yield from request.json()
+            client = self.data.client
+            for sha in data:
+                _LOGGER.info("Deleting: %s", sha)
+                client.delete(sha)
+        except ValueError:
+            return self.json_message('Bad item id', HTTP_BAD_REQUEST)
+
+
+class AsteriskMboxMP3View(HomeAssistantView):
+    """View to return an MP3."""
+
+    url = r"/api/asteriskmbox/mp3/{sha:[0-9a-f]+}"
+    name = "api:asteriskmbox:mp3"
+
+    def __init__(self, data):
+        """Initialize."""
+        self.data = data
+
+    @asyncio.coroutine
+    def get(self, request, sha):
+        """Retrieve Asterisk mp3."""
+        _LOGGER.info("Sending mp3 for %s", sha)
+
+        response = yield from self.handle(request, sha)
```

Comment:
I created the handle because I want to yield to it.  The callback to the asterisk server may take a significant time (hopefully not), and I didn't want to block other jobs while waiting for the reply.  I am not sure of a better way to address this.